fdbserver/MoveKeys.actor.cpp

```diff
@@ -1302,15 +1343,47 @@ ACTOR Future<Void> removeKeysFromFailedServer(Database cx,
 						continue;
 					}
 
+					TraceEvent("FailedServerRemoveBegin", serverID)
+					    .detail("Key", it.key)
+					    .detail("ValueSrc", describe(src))
+					    .detail("ValueDest", describe(dest));
+
 					// Update the vectors to remove failed server then set the value again
 					// Dest is usually empty, but keep this in case there is parallel data movement
 					src.erase(std::remove(src.begin(), src.end(), serverID), src.end());
 					dest.erase(std::remove(dest.begin(), dest.end(), serverID), dest.end());
-					TraceEvent(SevDebug, "FailedServerSetKey", serverID)
-					    .detail("Key", it.key)
-					    .detail("ValueSrc", describe(src))
-					    .detail("ValueDest", describe(dest));
-					tr.set(keyServersKey(it.key), keyServersValue(UIDtoTagMap, src, dest));
+					// If the last src server is to be removed, first check if there are dest servers who is
+					// hosting a read-write copy of the data, and move such dest servers to the src list.
+					if (src.empty() && !dest.empty()) {
+						std::vector<UID> newSources =
+						    wait(pickReadWriteServers(&tr, dest, KeyRangeRef(it.key, keyServers[i + 1].key)));
+						for (const UID& id : newSources) {
```

Comment:
Also just to confirm, should newSources contain full data from the old src?

Reply:
If there is at least one source server left for the team, we won't add a newSources, this happens only if the source list is going to be empty, after removing the failed server.